### PR TITLE
[BEAM-5105] Better parallelize BigQuery load jobs

### PR DESCRIPTION
**Pull request by reuvenlax**
Instead of blocking in every processElement, start the load jobs and wait for all of them in finishBundle.

### CODE REVIEW

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIOWriteTest.java [hunk 1]
```diff
@@ -757,7 +757,7 @@ public void testWriteUnknown() throws Exception {
                 .withoutValidation());
 
     thrown.expect(RuntimeException.class);
-    thrown.expectMessage("Failed to create load job");
+    thrown.expectMessage("Failed to create job");
```
Comment (chamikaramj) [inline sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIOWriteTest.java]: Can you add tests for some of the new utilities ? For example, PendingJobManager and PendingJob.

**Comment by aaltay** [comment]
What is the status of this PR?

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIOWriteTest.java (hunk 1 above)
Comment (reuvenlax) [reply sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIOWriteTest.java]: done

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryHelpers.java [hunk 2]
```diff
@@ -79,24 +86,202 @@ public RetryJobIdResult(RetryJobId jobId, boolean shouldRetry) {
...
+                BigQueryHelpers.getRetryJobId(currentJobId, projectId, bqLocation, jobService)
+                    .jobId;
+            LOG.info(
+                "Load job {} failed, {}: {}. Next job id {}",
+                oldJobId,
+                shouldRetry() ? "will retry" : "will not retry",
+                job.getStatus(),
+                currentJobId);
+            return false;
+          default:
```
[the remark this answers is not in the document]
Comment (reuvenlax) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryHelpers.java]: no

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryHelpers.java [hunk 3]
```diff
@@ -79,24 +86,202 @@ public RetryJobIdResult(RetryJobId jobId, boolean shouldRetry) {
...
+            // Job has completed successfully.
+            Exception e = jobInfo.onSuccess.apply(null);
+            if (e != null) {
+              throw e;
+            }
+          } else {
+            // Job failed, start it again. If it has hit the maximum number of retries then runJob
+            // will throw an exception.
+            jobInfo.pendingJob.runJob();
+            retryJobs.add(jobInfo);
```
[the remark this answers is not in the document]
Comment (reuvenlax) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryHelpers.java]: good catch. fixed

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/WriteTables.java [hunk 4]
```diff
@@ -166,22 +184,64 @@ public void processElement(ProcessContext c) throws Exception {
...
+                        pendingJob.window.maxTimestamp(),
+                        pendingJob.window);
+                    for (String file : pendingJob.partitionFiles) {
+                      c.output(
+                          temporaryFilesTag,
+                          file,
+                          pendingJob.window.maxTimestamp(),
+                          pendingJob.window);
+                    }
+                    return null;
```
[the remark this answers is not in the document]
Comment (reuvenlax) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/WriteTables.java]: function needs to return something.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/WriteRename.java [hunk 5]
```diff
@@ -88,148 +102,128 @@ public void processElement(ProcessContext c) throws Exception {
...
+            try {
+              if (pendingJob.tableDestination.getTableDescription() != null) {
+                TableReference ref = pendingJob.tableDestination.getTableReference();
+                datasetService.patchTableDescription(
+                    ref.clone()
+                        .setTableId(BigQueryHelpers.stripPartitionDecorator(ref.getTableId())),
+                    pendingJob.tableDestination.getTableDescription());
+              }
+              removeTemporaryTables(datasetService, pendingJob.tempTables);
+              return null;
```
[the remark this answers is not in the document]
Comment (reuvenlax) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/WriteRename.java]: fitto

**Comment by chamikaramj** [comment]
Doesn't look like we have large scale jobs in Beam. I think we have large scale jobs internally.

**Pull request merged**